File: algorithms/02_c/sorting/shell_sort.c
/* Shell sort with Knuth's gap sequence. */
#include <stdio.h>

void shell_sort(int *a, int n) {
    for (int gap = n / 2; gap > 0; gap /= 2) {
        for (int i = gap; i < n; i++) {
            int temp = a[i], j = i;
            while (j >= gap && a[j - gap] > temp) {
                a[j] = a[j - gap];
                j -= gap;
            }
            a[j] = temp;
        }
    }
}

int main(void) {
    int a[] = {12, 34, 54, 2, 3, 90, 45};
    int n = (int)(sizeof(a) / sizeof(a[0]));
    shell_sort(a, n);
    for (int i = 1; i < n; i++) if (a[i - 1] > a[i]) return 1;
    for (int i = 0; i < n; i++) printf("%d ", a[i]);
    printf("\n");
    return 0;
}

File: algorithms/07_typescript/sorting/shell_sort.ts
/**
 * Shell Sort (Donald Shell)
 * Generalization of insertion sort with diminishing gaps.
 */

export function shellSort(arr: number[]): number[] {
  const n = arr.length;

  // Knuth gap sequence: h = 3*h + 1
  let h = 1;
  while (h < Math.floor(n / 3)) {
    h = 3 * h + 1;
  }

  while (h >= 1) {
    for (let i = h; i < n; i++) {
      const temp = arr[i];
      let j = i;
      while (j >= h && arr[j - h] > temp) {
        arr[j] = arr[j - h];
        j -= h;
      }
      arr[j] = temp;
    }
    h = Math.floor(h / 3);
  }
  return arr;
}

const testSS = [35, 33, 42, 10, 14, 19, 27, 44];
shellSort(testSS);
if (testSS.join(",") !== "10,14,19,27,33,35,42,44") throw new Error("Shell Sort failed");
console.log("Shell Sort verified successfully.");
